##Library innitialization
```{r}
library(TCGAbiolinks)
library(SummarizedExperiment)
library(data.table)
library(dplyr)
library(DT)
library(survival)


#library("categoryCompare")
#library("GO.db")
#library("KEGG.db")

```


##Download data, prepare data into matrix, normalize, filter
```{r}
CancerProject <- "TCGA-LUAD"

query <- GDCquery(project = CancerProject,
                  data.category = "Transcriptome Profiling",
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "HTSeq - FPKM")

samplesDown <- getResults(query,cols=c("cases"))

dataSmTP <- TCGAquery_SampleTypes(barcode = samplesDown,
                                  typesample = "TP")

#dataSmNT <- TCGAquery_SampleTypes(barcode = samplesDown,typesample = "NT")

#dataSmTPNT<-TCGAquery_MatchedCoupledSampleTypes(barcode = samplesDown,c("NT","TP"))

#dataSmTPNT<-TCGAquery_MatchedCoupledSampleTypes(barcode = samplesDown,c("NT","TP"))



queryDown <- GDCquery(project = CancerProject, 
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification", 
                      workflow.type = "HTSeq - FPKM", 
                      barcode = dataSmTP)

GDCdownload(query = queryDown,directory = "GDCdata")
#GDCdownload(query = queryDown,method = "client",chunks.per.download = 10)

# get indexed clinical data
dataClin <- GDCquery_clinic(project = CancerProject, "Clinical")

# get subtype information
dataSubt <- TCGAquery_subtype(tumor = "LUAD")

dataPrep <- GDCprepare(query = queryDown,
                       directory = "GDCdata",
                       save = TRUE, 
                       save.filename="dataPrep.rda")

#get subtype information
#dataSubt <- TCGAquery_subtype(tumor = "BRCA")

#dataPrep <- subset(dataPrep, select = colData(dataPrep)$patient %in% dataSubt$patient)

#changes data from RangedSummarizedExperiment to matrix
dataPrep2 <- TCGAanalyze_Preprocessing(object = dataPrep, 
                                      cor.cut = 0.0,
                                      datatype = "HTSeq - FPKM")                      

#dataNorm <- TCGAanalyze_Normalization(tabDF = dataPrep2,
#                                      geneInfo = geneInfoHT,
#                                      method = "gcContent")

#boxplot(dataPrep, outline = FALSE)

#dataFilt <- TCGAanalyze_Filtering(tabDF = dataPrep2, method = "quantile", qnt.cut =  0.0)  

#dataFilt1 <- TCGAanalyze_Filtering(tabDF = dataNorm,method = "varFilter")
#dataFilt2 <- TCGAanalyze_Filtering(tabDF = dataFilt1,method = "filter1")
#dataFilt <- TCGAanalyze_Filtering(tabDF = dataFilt2,method = "filter2")

#Filter out genes that have zeros in more than 50% of cases
numGenes<-length(dataPrep2[,1])
numCases<-length(dataPrep2[1,])
status<-0
for(ii in 1:numGenes){
  if(length(which(dataPrep2[ii,]<0.000000001))>0.5*numCases)
    status[ii] <- 1
  else
    status[ii] <- 0
}
dataPrep3<-dataPrep2[which(status==0),]
dim(dataPrep2)
dim(dataPrep3)
#write.table(scale(t(log2(dataPrep3+1)), center = TRUE, scale = TRUE), file = "out/LUADGeneTable.csv", sep = ",", col.names = NA, qmethod = "double")

#--------------Find MAF gene-------------------------------------
grep("ENSG00000178573",row.names(dataPrep3))
#dataNorm[grep("ENSG00000178573",row.names(dataNorm)),]

```

##Filter and keep only differentially expressed genes
```{r}
dataDEGs<-read.table(file = "../tumorNormalProject/out/DiffExpLUADTableLFC2.csv", sep = ",", header=TRUE, row.names=1)
dim(dataDEGs)
dim(dataPrep3)
dataPrep3DE<-dataPrep3[rownames(dataPrep3) %in% rownames(dataDEGs),]
dim(dataPrep3DE)
dim(dataDEGs)

grep("ENSG00000178573",row.names(dataPrep3DE))

write.table(scale(t(log2(dataPrep3DE+1)), center = TRUE, scale = TRUE), file = "out/LUADGeneTableDELFC2.csv", sep = ",", col.names = NA, qmethod = "double")

```
